File: native/core/overlay/render.cpp
#include "overlay/render.hpp"

#include <QBrush>
#include <QColor>
#include <QFont>
#include <QFontDatabase>
#include <QFontMetricsF>
#include <QImage>
#include <QLinearGradient>
#include <QPainter>
#include <QPainterPath>
#include <QRadialGradient>
#include <QRectF>
#include <QSize>
#include <QString>
#include <QStringList>
#include <QTransform>

#include <algorithm>
#include <cmath>
#include <map>
#include <mutex>
#include <string>
#include <utility>
#include <vector>

namespace sstvae::overlay {

namespace {

// ---------------------------------------------------------------------
// Pictures in and out

QImage to_qimage(const images::Picture& p) {
    if (p.empty()) return QImage();
    // The Picture's rows are tightly packed; QImage would otherwise
    // assume a 4-byte-aligned stride and shear the picture. copy()
    // because the wrapper does not own the bytes.
    const QImage view(p.rgb.data(), p.width, p.height, p.width * 3,
                      QImage::Format_RGB888);
    return view.copy().convertToFormat(QImage::Format_ARGB32_Premultiplied);
}

images::Picture from_qimage(const QImage& image) {
    const QImage rgb = image.convertToFormat(QImage::Format_RGB888);
    images::Picture out(rgb.width(), rgb.height());
    for (int y = 0; y < rgb.height(); ++y) {
        std::copy_n(rgb.constScanLine(y), static_cast<std::size_t>(rgb.width()) * 3,
                    out.rgb.data() + static_cast<std::size_t>(y) * rgb.width() * 3);
    }
    return out;
}

QColor color_of(const std::string& text, const QColor& fallback) {
    const QColor c(QString::fromStdString(text));
    return c.isValid() ? c : fallback;
}

// ---------------------------------------------------------------------
// Fonts
//
// The document names a font by *path*, because that is what PIL takes
// and what a saved template carries. Qt matches by family, so a path
// has to be registered with the font database first and the family it
// contributed read back out. Cached: registering is file I/O, and a
// text item is re-measured on every drag in the editor.

QString family_for(const std::string& path) {
    if (path.empty()) return QString();
    static std::mutex mutex;
    static std::map<std::string, QString> cache;
    const std::lock_guard<std::mutex> lock(mutex);
    const auto it = cache.find(path);
    if (it != cache.end()) return it->second;

    QString family;
    const int id = QFontDatabase::addApplicationFont(QString::fromStdString(path));
    if (id != -1) {
        const QStringList families = QFontDatabase::applicationFontFamilies(id);
        if (!families.isEmpty()) family = families.front();
    }
    cache.emplace(path, family);
    return family;
}

// The document's four generic keywords. Qt's fontconfig backend
// understands them as family names, but the other platforms' databases
// do not, so the style hint rides along: it is what makes "monospace"
// mean a fixed-pitch face on a machine where no family is called that.
// Anything else is a real family name and is passed through untouched.
void apply_family_request(QFont& font, const std::string& request) {
    const QString name = QString::fromStdString(request);
    struct Generic {
        const char* keyword;
        QFont::StyleHint hint;
    };
    static constexpr Generic GENERICS[] = {
        {"sans-serif", QFont::SansSerif},
        {"serif", QFont::Serif},
        {"monospace", QFont::Monospace},
        {"cursive", QFont::Cursive},
    };
    for (const Generic& g : GENERICS) {
        if (name.compare(QLatin1String(g.keyword), Qt::CaseInsensitive) == 0) {
            font.setFamily(name);
            font.setStyleHint(g.hint);
            return;
        }
    }
    font.setFamily(name);
}

QFont font_for(const TextItem& item, int size_px) {
    QFont font;
    // `font` (a path) wins over `font_family`: a template that ships its
    // own face is naming the exact file it needs. The family request is
    // the fallback for when there is no path *or* the path yielded no
    // family, so an unreadable file degrades to the operator's choice of
    // face rather than to the default.
    const QString family = family_for(item.font);
    if (!family.isEmpty()) font.setFamily(family);
    else if (!item.font_family.empty()) apply_family_request(font, item.font_family);
    font.setBold(item.bold);
    font.setItalic(item.italic);
    // setPixelSize, not setPointSize: the document sizes text as a
    // fraction of canvas height, so the answer must not depend on the
    // DPI of whatever screen happens to be attached.
    font.setPixelSize(std::max(1, size_px));
    return font;
}

// ---------------------------------------------------------------------
// Text layout
//
// PIL's two-letter anchor, kept because it is what the document says
// and what the reference's saved files contain. First letter is
// horizontal (l/m/r), second vertical: 'a' ascender, 't' top, 'm'
// middle, 's' baseline, 'b' bottom, 'd' descender.

struct TextLayout {
    QStringList lines;
    double line_height = 0.0;  // baseline to baseline
    double width = 0.0;
    double height = 0.0;
    double ascent = 0.0;
    double left = 0.0;  // top-left of the block, anchor applied
    double top = 0.0;
};

TextLayout layout_text(const TextItem& item, const QFont& font, int size_px,
                       double x, double y) {
    const QFontMetricsF fm(font);
    TextLayout out;
    out.lines = QString::fromStdString(item.text).split(QLatin1Char('\n'));
    if (out.lines.isEmpty()) out.lines << QString();

    const double spacing = std::max(0.0, item.line_spacing * size_px);
    out.ascent = fm.ascent();
    out.line_height = fm.height() + spacing;
    for (const QString& line : out.lines)
        out.width = std::max(out.width, fm.horizontalAdvance(line));
    out.height = fm.height() +
                 out.line_height * static_cast<double>(out.lines.size() - 1);

    const char h = item.anchor.empty() ? 'l' : item.anchor[0];
    const char v = item.anchor.size() > 1 ? item.anchor[1] : 'a';

    out.left = x;
    if (h == 'm') out.left = x - out.width / 2.0;
    else if (h == 'r') out.left = x - out.width;

    switch (v) {
        case 'm': out.top = y - out.height / 2.0; break;
        case 's': out.top = y - out.ascent; break;
        case 'b':
        case 'd': out.top = y - out.height; break;
        default: out.top = y; break;  // 'a' ascender, 't' top
    }
    return out;
}

// Where an underline sits, from the font's own metrics. Shared by the
// drawing and by `item_bbox`, so a selection handle cannot stop short
// of a line the renderer draws.
//
// Never thinner than a pixel: `lineWidth()` is 0 or fractional for some
// faces at small sizes, and an underline that vanishes there would make
// the toggle look broken.
QRectF underline_rect(const QFontMetricsF& fm, double x, double baseline,
                      double width) {
    return QRectF(x, baseline + fm.underlinePos(), width,
                  std::max(1.0, fm.lineWidth()));
}

// The block as two paths, so the stroke is drawn under *all* the glyphs
// before any of them is filled. Stroking and filling line by line would
// let a descender's outline cut across the line below it.
//
// **The underline is its own path rather than a rect appended to the
// glyphs'.** `QPainterPath::addText` adds outlines only -- a font's
// underline is a decoration Qt draws separately, so
// `QFont::setUnderline` renders nothing through this path -- and a rect
// appended to the same path fights it over the fill rule: under
// odd-even a descender crossing the line is punched out of it, and
// under winding a contour running the other way cancels it. Two paths,
// both stroked before either is filled, is what a union would look like
// without asking Qt to compute one.
struct TextShape {
    QPainterPath glyphs;
    QPainterPath underline;
};

TextShape text_shape(const TextItem& item, const QFont& font,
                     const TextLayout& layout) {
    const QFontMetricsF fm(font);
    TextShape shape;
    for (int i = 0; i < layout.lines.size(); ++i) {
        const QString& line = layout.lines[i];
        if (line.isEmpty()) continue;
        const double advance = fm.horizontalAdvance(line);
        double x = layout.left;
        if (item.align == "center") x += (layout.width - advance) / 2.0;
        else if (item.align == "right") x += layout.width - advance;
        const double baseline =
            layout.top + layout.ascent + layout.line_height * i;
        shape.glyphs.addText(QPointF(x, baseline), font, line);
        if (item.underline)
            shape.underline.addRect(underline_rect(fm, x, baseline, advance));
    }
    return shape;
}

// The fill. `color` is the first stop in every mode, which is what
// keeps a version-1 document identical: a solid item takes the exact
// path it always did.
//
// Built over the layout box rather than the ink, so the ramp does not
// shift as the text is edited, and in the painter's current space --
// already translated and rotated for the item -- so the gradient turns
// with the text instead of sliding across it. An unrecognised mode is
// solid, per the document.
QBrush fill_brush(const TextItem& item, const TextLayout& layout) {
    const QColor from = color_of(item.color, Qt::white);
    const bool linear = item.fill_mode == "linear";
    const bool radial = item.fill_mode == "radial";
    if (!linear && !radial) return QBrush(from);

    const QColor to = color_of(item.color2, from);
    const QPointF centre(layout.left + layout.width / 2.0,
                         layout.top + layout.height / 2.0);
    if (radial) {
        const double radius =
            std::max(1.0, std::hypot(layout.width, layout.height) / 2.0);
        QRadialGradient g(centre, radius);
        g.setColorAt(0.0, from);
        g.setColorAt(1.0, to);
        return QBrush(g);
    }

    // Clockwise on screen (y grows downward), so 0 runs left to right
    // and 90 top to bottom. The half-length is the box's extent along
    // that direction, which puts the end stops exactly on the corners --
    // any shorter and a diagonal ramp never reaches its far colour.
    // Not M_PI: MSVC only defines it behind _USE_MATH_DEFINES.
    constexpr double PI = 3.14159265358979323846;
    const double a = item.fill_angle * PI / 180.0;
    const double dx = std::cos(a);
    const double dy = std::sin(a);
    const double half =
        (std::abs(layout.width * dx) + std::abs(layout.height * dy)) / 2.0;
    QLinearGradient g(centre - QPointF(dx, dy) * half,
                      centre + QPointF(dx, dy) * half);
    g.setColorAt(0.0, from);
    g.setColorAt(1.0, to);
    return QBrush(g);
}

void draw_text(QPainter& painter, const TextItem& item, int canvas_w,
               int canvas_h) {
    if (item.text.empty()) return;
    const int size_px = static_cast<int>(std::lround(item.size * canvas_h));
    const QFont font = font_for(item, size_px);
    const double stroke =
        std::max(0.0, item.stroke_width * item.size * canvas_h);
    const double x = std::lround(item.x * canvas_w);
    const double y = std::lround(item.y * canvas_h);

    painter.save();
    if (item.rotation != 0.0) {
        // About the anchor point, which is the point the document
        // actually pins. (The reference rotates a padded layer and
        // composites it at the anchor, which is close but not the same;
        // this is the version an editor can show a handle for.) Negated
        // because the document's angle is counter-clockwise, as PIL's
        // is, and QTransform::rotate turns the other way.
        painter.translate(x, y);
        painter.rotate(-item.rotation);
        painter.translate(-x, -y);
    }

    const TextLayout layout = layout_text(item, font, size_px, x, y);
    const TextShape shape = text_shape(item, font, layout);
    if (stroke > 0.0) {
        QPen pen(color_of(item.stroke_color, Qt::black));
        // PIL's stroke_width is a radius, drawn outside the glyph; a
        // Qt pen straddles the path, so half of a width-2w pen lands
        // outside. Same visual weight rather than a coincidence.
        pen.setWidthF(stroke * 2.0);
        pen.setJoinStyle(Qt::RoundJoin);
        painter.strokePath(shape.glyphs, pen);
        if (item.underline) painter.strokePath(shape.underline, pen);
    }
    const QBrush brush = fill_brush(item, layout);
    painter.fillPath(shape.glyphs, brush);
    if (item.underline) painter.fillPath(shape.underline, brush);
    painter.restore();
}

// ---------------------------------------------------------------------
// Image insets

// A file-backed inset, decoded once.
//
// **Keyed on the path and never invalidated**, which is the whole
// contract: a document names an inset by path, so the only thing that
// changes which pixels an item shows is the path changing. Watching the
// file for content changes is deliberately not attempted -- it would
// mean a stat on every call, on a path that runs on every mouse move,
// to catch a case (the operator overwriting a file the composition
// already refers to, in place, mid-session) that costs nothing to fix
// by re-adding the inset.
//
// The decode used to happen on every call, and `item_bbox` is a caller:
// `OverlayEditor::hit_test` runs it per item on **every mouse move over
// the canvas**, and `paintEvent` runs it again. So merely moving the
// pointer across the composer re-read and re-decoded every inset from
// disk.
//
// A failed load is cached too. The alternative is retrying a missing
// file at mouse-move rate, which is the same pathology with syscalls
// instead of a decode.
//
// Returned by value: `QImage` is copy-on-write, so this is a refcount
// bump, and it means no caller holds a reference into a cache another
// thread could evict.
QImage cached_file_image(const std::string& path) {
    // Small and FIFO-bounded rather than unbounded. A source photograph
    // is not small -- 4000x3000 is 48 MB as ARGB32 -- and this is a
    // process-lifetime cache in an application that runs for days, so
    // an operator cycling through a folder of pictures must not be able
    // to grow it without limit. Eight is far more than any composition
    // uses at once.
    constexpr std::size_t CAPACITY = 8;
    static std::mutex mutex;
    static std::vector<std::pair<std::string, QImage>> cache;

    const std::lock_guard<std::mutex> lock(mutex);
    for (const auto& entry : cache) {
        if (entry.first == path) return entry.second;
    }

    QImage loaded;
    if (loaded.load(QString::fromStdString(path))) {
        loaded = loaded.convertToFormat(QImage::Format_ARGB32_Premultiplied);
    } else {
        loaded = QImage();
    }
    if (cache.size() >= CAPACITY) cache.erase(cache.begin());
    cache.emplace_back(path, loaded);
    return loaded;
}

QImage resolve_source(const ImageItem& item, const images::Picture* last_rx) {
    if (item.source == SOURCE_LAST_RX) {
        return last_rx != nullptr ? to_qimage(*last_rx) : QImage();
    }
    if (item.source.empty()) return QImage();
    return cached_file_image(item.source);
}

// The source's dimensions, without materializing it.
//
// `item_bbox` wants nothing from the source but its aspect ratio, and
// it is the call that runs on every mouse move. For a "last_rx" item
// the answer is already on the `Picture`, so going through
// `resolve_source` converted a 640x480 reception into a 1.2 MB ARGB32
// `QImage` in order to read two integers back off it.
QSize source_size(const ImageItem& item, const images::Picture* last_rx) {
    if (item.source == SOURCE_LAST_RX) {
        if (last_rx == nullptr || last_rx->empty()) return QSize();
        return QSize(last_rx->width, last_rx->height);
    }
    if (item.source.empty()) return QSize();
    return cached_file_image(item.source).size();
}

// The inset at its drawn size, border included; the anchor is applied
// by the caller, which is also what item_bbox does.
QSize inset_size(const ImageItem& item, int canvas_w, double aspect) {
    const int w = std::max(1, static_cast<int>(std::lround(item.width * canvas_w)));
    const int h = std::max(1, static_cast<int>(std::lround(w * aspect)));
    const int border =
        std::max(0, static_cast<int>(std::lround(item.border * canvas_w)));
    return QSize(w + 2 * border, h + 2 * border);
}

void apply_anchor(const std::string& anchor, int w, int h, int& x, int& y) {
    const char horizontal = anchor.empty() ? 'l' : anchor[0];
    const char vertical = anchor.size() > 1 ? anchor[1] : 'a';
    if (horizontal == 'm') x -= w / 2;
    else if (horizontal == 'r') x -= w;
    if (vertical == 'm') y -= h / 2;
    else if (vertical == 'b' || vertical == 'd') y -= h;
}

void draw_image(QPainter& painter, const ImageItem& item, int canvas_w,
                int canvas_h, const images::Picture* last_rx) {
    const QImage src = resolve_source(item, last_rx);
    if (src.isNull()) return;

    const int border =
        std::max(0, static_cast<int>(std::lround(item.border * canvas_w)));
    const int target_w =
        std::max(1, static_cast<int>(std::lround(item.width * canvas_w)));
    const int target_h = std::max(
        1, static_cast<int>(std::lround(static_cast<double>(target_w) *
                                        src.height() / src.width())));

    QImage inset(target_w + 2 * border, target_h + 2 * border,
                 QImage::Format_ARGB32_Premultiplied);
    inset.fill(border > 0 ? color_of(item.border_color, Qt::white)
                          : QColor(Qt::transparent));
    {
        QPainter inset_painter(&inset);
        inset_painter.setRenderHint(QPainter::SmoothPixmapTransform, true);
        inset_painter.drawImage(QRect(border, border, target_w, target_h), src);
    }

    int x = static_cast<int>(std::lround(item.x * canvas_w));
    int y = static_cast<int>(std::lround(item.y * canvas_h));
    apply_anchor(item.anchor, inset.width(), inset.height(), x, y);

    painter.save();
    painter.setOpacity(std::clamp(item.opacity, 0.0, 1.0));
    if (item.rotation != 0.0) {
        const QPointF centre(x + inset.width() / 2.0, y + inset.height() / 2.0);
        painter.translate(centre);
        painter.rotate(-item.rotation);
        painter.translate(-centre);
    }
    painter.drawImage(QPoint(x, y), inset);
    painter.restore();
}

}  // namespace

Bbox item_bbox(int canvas_w, int canvas_h, const Item& item,
               const images::Picture* last_rx) {
    if (const TextItem* text = std::get_if<TextItem>(&item)) {
        const int size_px =
            static_cast<int>(std::lround(text->size * canvas_h));
        const QFont font = font_for(*text, size_px);
        const double stroke =
            std::max(0.0, text->stroke_width * text->size * canvas_h);
        const double x = std::lround(text->x * canvas_w);
        const double y = std::lround(text->y * canvas_h);
        // A space when empty, so an item being typed into still has a
        // handle to select rather than collapsing to nothing.
        TextItem measured = *text;
        if (measured.text.empty()) measured.text = " ";
        const TextLayout layout = layout_text(measured, font, size_px, x, y);
        // An underline can sit below the font's descent, and the handle
        // must not clip a line the renderer draws.
        double bottom = layout.top + layout.height;
        if (text->underline) {
            const double baseline = layout.top + layout.ascent +
                                    layout.line_height * (layout.lines.size() - 1);
            const QRectF line =
                underline_rect(QFontMetricsF(font), layout.left, baseline, layout.width);
            bottom = std::max(bottom, line.bottom());
        }
        return Bbox{static_cast<int>(std::lround(layout.left - stroke)),
                    static_cast<int>(std::lround(layout.top - stroke)),
                    std::max(1, static_cast<int>(
                                    std::lround(layout.width + 2 * stroke))),
                    std::max(1, static_cast<int>(
                                    std::lround(bottom - layout.top + 2 * stroke)))};
    }

    const ImageItem& image = std::get<ImageItem>(item);
    // Dimensions only -- see `source_size`. This is the mouse-move path.
    const QSize src = source_size(image, last_rx);
    const double aspect =
        src.isEmpty() ? 0.75
                      : static_cast<double>(src.height()) / src.width();
    const QSize size = inset_size(image, canvas_w, aspect);
    int x = static_cast<int>(std::lround(image.x * canvas_w));
    int y = static_cast<int>(std::lround(image.y * canvas_h));
    apply_anchor(image.anchor, size.width(), size.height(), x, y);
    return Bbox{x, y, size.width(), size.height()};
}

images::Picture render(const images::Picture& base, const Doc& doc,
                       const images::Picture* last_rx) {
    if (base.empty()) return base;
    QImage canvas = to_qimage(base);
    {
        QPainter painter(&canvas);
        painter.setRenderHint(QPainter::Antialiasing, true);
        painter.setRenderHint(QPainter::TextAntialiasing, true);
        painter.setRenderHint(QPainter::SmoothPixmapTransform, true);
        for (const Item& item : doc.items) {
            if (const TextItem* text = std::get_if<TextItem>(&item)) {
                draw_text(painter, *text, base.width, base.height);
            } else if (const ImageItem* image = std::get_if<ImageItem>(&item)) {
                draw_image(painter, *image, base.width, base.height, last_rx);
            }
        }
    }
    return from_qimage(canvas);
}

}  // namespace sstvae::overlay

File: sstvae/overlay/render.py
"""Rendering an `OverlayDoc` onto a picture.

Pure PIL, no Qt: the GUI editor draws its own live preview with Qt
items, but what actually gets transmitted is rendered here, so the
result is identical whether it came from the editor, a future saved
template, or a command-line `--overlay doc.json`.

Note the codec is trained for exactly this kind of content -- see the
burned-in text augmentation in `sstvae/data.py` and its comment on why
the training text is deliberately unstructured. Composition happens
*before* encoding, so the overlay is part of the picture the network
codes, not something laid on afterwards.

**The document version 2 style fields are deliberately not drawn here**
-- bold, italic, underline, `font_family` and the gradient fills. That
is not an oversight: the application that grew the style palette is the
C++ one, this module is the reference for the *document* plus what the
Python tests render, and duplicating a gradient in PIL would be a second
implementation to keep in step for no caller. The defaults in `model.py`
are what make the omission safe -- a document written without touching
the palette renders here exactly as it always did.
"""

import os
from functools import lru_cache

from PIL import Image, ImageDraw, ImageFont

from .model import ImageItem, OverlayDoc, SOURCE_LAST_RX, TextItem

# Same font search the training overlays use, so the GUI's default face
# matches what the model was trained on rather than being an arbitrary
# second choice.
from ..images import AVAILABLE_FONTS, open_image


@lru_cache(maxsize=64)
def _load_font(path: str | None, size: int):
    size = max(1, int(size))
    candidates = [path] if path else []
    candidates += list(AVAILABLE_FONTS)
    for p in candidates:
        if p and os.path.exists(p):
            try:
                return ImageFont.truetype(p, size)
            except OSError:
                continue
    try:
        return ImageFont.load_default(size=size)
    except TypeError:  # Pillow < 10.1 has a bitmap-only default
        return ImageFont.load_default()


def _text_draw_kwargs(item, size_px: int) -> dict:
    """The PIL keyword arguments for drawing (or measuring) a text item.

    In one place so `item_bbox` and `_render_text` cannot disagree about
    geometry -- the editor's selection handle comes from the former and
    the picture from the latter. PIL applies `align` and `spacing` only
    when the string actually spans lines.
    """
    return {
        "anchor": item.anchor,
        "align": item.align,
        "spacing": max(0, round(item.line_spacing * size_px)),
    }


def _resolve_source(source: str, last_rx: Image.Image | None) -> Image.Image | None:
    """Late binding: turn an `ImageItem.source` reference into pixels.

    Returning None (rather than raising) for a missing source is
    deliberate -- a template referring to the last received image is
    perfectly valid on a session where nothing has been received yet,
    and should simply draw nothing.
    """
    if source == SOURCE_LAST_RX:
        return last_rx
    if not source or not os.path.exists(source):
        return None
    try:
        # Upright, like the main picture -- an inset is usually a
        # photograph too, and one of the two arriving sideways would be
        # the more confusing outcome.
        return open_image(source).convert("RGB")
    except (OSError, ValueError):
        # ValueError is `open_image`'s size refusal. An inset that is too
        # large to open draws nothing, like every other unusable source
        # here -- refusing to render the whole composition over one
        # decorative element would be the worse failure.
        return None


def _render_text(canvas: Image.Image, item: TextItem) -> None:
    if not item.text:
        return
    w, h = canvas.size
    size_px = round(item.size * h)
    font = _load_font(item.font, size_px)
    stroke = max(0, round(item.stroke_width * item.size * h))
    x, y = round(item.x * w), round(item.y * h)
    kw = _text_draw_kwargs(item, size_px)

    if not item.rotation:
        ImageDraw.Draw(canvas).text(
            (x, y), item.text, font=font, fill=item.color,
            stroke_width=stroke, stroke_fill=item.stroke_color, **kw,
        )
        return

    # Rotated text needs its own layer: PIL cannot rotate a draw call.
    pad = stroke * 2 + 4
    box = ImageDraw.Draw(Image.new("RGBA", (1, 1))).textbbox(
        (0, 0), item.text, font=font, stroke_width=stroke, **kw
    )
    layer = Image.new(
        "RGBA", (box[2] - box[0] + 2 * pad, box[3] - box[1] + 2 * pad), (0, 0, 0, 0)
    )
    ImageDraw.Draw(layer).text(
        (pad - box[0], pad - box[1]), item.text, font=font, fill=item.color,
        stroke_width=stroke, stroke_fill=item.stroke_color, **kw,
    )
    layer = layer.rotate(item.rotation, resample=Image.BICUBIC, expand=True)
    canvas.alpha_composite(layer, (x - pad, y - pad))


def _render_image(canvas: Image.Image, item: ImageItem,
                  last_rx: Image.Image | None) -> None:
    src = _resolve_source(item.source, last_rx)
    if src is None:
        return
    w, h = canvas.size
    target_w = max(1, round(item.width * w))
    target_h = max(1, round(target_w * src.height / src.width))
    inset = src.convert("RGBA").resize((target_w, target_h), Image.LANCZOS)

    border = max(0, round(item.border * w))
    if border:
        framed = Image.new(
            "RGBA",
            (target_w + 2 * border, target_h + 2 * border),
            item.border_color,
        )
        framed.paste(inset, (border, border))
        inset = framed

    if item.opacity < 1.0:
        alpha = inset.getchannel("A").point(
            lambda a: round(a * max(0.0, min(1.0, item.opacity)))
        )
        inset.putalpha(alpha)

    if item.rotation:
        inset = inset.rotate(item.rotation, resample=Image.BICUBIC, expand=True)

    x, y = round(item.x * w), round(item.y * h)
    if item.anchor.startswith("m"):
        x -= inset.width // 2
    elif item.anchor.startswith("r"):
        x -= inset.width
    if item.anchor.endswith("m"):
        y -= inset.height // 2
    elif item.anchor.endswith("b") or item.anchor.endswith("d"):
        y -= inset.height
    canvas.alpha_composite(inset, (x, y))


def item_bbox(canvas_size: tuple[int, int], item,
              last_rx: Image.Image | None = None) -> tuple[int, int, int, int]:
    """Pixel (x, y, w, h) an item occupies on a canvas of `canvas_size`.

    Lives here rather than in the editor so the on-screen selection
    handles are positioned by the same geometry that draws the item --
    otherwise the two drift apart and the handle stops matching what the
    operator sees.
    """
    w, h = canvas_size
    x, y = round(item.x * w), round(item.y * h)

    if isinstance(item, TextItem):
        size_px = round(item.size * h)
        font = _load_font(item.font, size_px)
        stroke = max(0, round(item.stroke_width * item.size * h))
        box = ImageDraw.Draw(Image.new("RGB", (1, 1))).textbbox(
            (x, y), item.text or " ", font=font, stroke_width=stroke,
            **_text_draw_kwargs(item, size_px),
        )
        return box[0], box[1], max(1, box[2] - box[0]), max(1, box[3] - box[1])

    src = _resolve_source(item.source, last_rx)
    aspect = (src.height / src.width) if src else 0.75
    iw = max(1, round(item.width * w))
    ih = max(1, round(iw * aspect))
    border = max(0, round(item.border * w))
    iw += 2 * border
    ih += 2 * border
    if item.anchor.startswith("m"):
        x -= iw // 2
    elif item.anchor.startswith("r"):
        x -= iw
    if item.anchor.endswith("m"):
        y -= ih // 2
    elif item.anchor.endswith(("b", "d")):
        y -= ih
    return x, y, iw, ih


def render(base: Image.Image, doc: OverlayDoc,
           last_rx: Image.Image | None = None) -> Image.Image:
    """Draw `doc` over `base` and return a new RGB image.

    `base` is used as-is; the caller is responsible for having framed it
    to the transmit size (`sstvae.images.fit_image`), because the
    document's normalized coordinates are relative to whatever it is
    given. `last_rx` supplies the pixels for any item whose source is
    `"last_rx"`.
    """
    canvas = base.convert("RGBA")
    for item in doc.items:
        if isinstance(item, TextItem):
            _render_text(canvas, item)
        elif isinstance(item, ImageItem):
            _render_image(canvas, item, last_rx)
    return canvas.convert("RGB")
